source/module_cell/print_cif.cpp

```diff
@@ -0,0 +1,169 @@
+#include "module_cell/unitcell.h"
+#include <cmath>   // for std::pow and std::round
+
+// Function to calculate the greatest common divisor
+long long gcd_func(long long a, long long b) {
+    return b == 0 ? a : gcd_func(b, a % b);
+}
+
+std::string floatToFraction(double number, double accuracy = 1e-9) {
+    // Determine the sign of the number
+    int sign = number < 0 ? -1 : 1;
+    number = std::abs(number);
+
+    // The maximum denominator is set to a large number to ensure precision
+    const long long max_denominator = 1000000;
+
+    // Start with a denominator of 1
+    long long denominator = 1;
+    long long numerator = std::round(number * denominator);
+
+    // Keep increasing the denominator to find a better approximation
+    double error;
+    do {
+        denominator *= 10;
+        numerator = std::round(number * denominator);
+        error = std::abs(number - (double)numerator / denominator);
+    } while (error > accuracy && denominator <= max_denominator);
+
+    // Simplify the fraction by dividing both numerator and denominator by their GCD
+    long long gcd = gcd_func(numerator, denominator);
+    numerator /= gcd;
+    denominator /= gcd;
+
+    // Apply the sign to the numerator
+    numerator *= sign;
+
+    if (denominator == 1) {
+        return std::to_string(numerator);
+    } else {
+        return std::to_string(numerator) + "/" + std::to_string(denominator);
+    }
+}
+
+void UnitCell::print_cell_cif(const std::string& fn) const
```

Comment:
I would suggest you write one or two functions, the kernel one can be kept as elemental as possible, and can be static function then all other classes/modules can use this function without any barrier. The param list of the kernel one can be:
```c++
// unitcell.h
static void cal_abc_angles_from_latvec(const ModuleBase::Vector3<double>& veca, const ModuleBase::Vector3<double>& vecb, const ModuleBase::Vector3<double>& vecc, double& a, double& b, double& c, double& alpha, double& beta, double& gamma); // then unittest can be easily equipped and can be use anywhere in ABACUS
// also a revert version
static void cal_latvec_from_abc_angles(const double&, const double&, const double&, const double&, const double&, const double&, const ModuleBase::Vector3<double>&, const ModuleBase::Vector3<double>&, const ModuleBase::Vector3<double>&); // this function can be used to cross-validate the above one
// the cif also need the cell volume, this can be easily done with mixed product of veca, vecb and vecc
static double cal_lattice_vol(const ModuleBase::Vector3<double>&, const ModuleBase::Vector3<double>&, const ModuleBase::Vector3<double>&);
// you can also expose an interface to the lattice parameter abc_angles
static double cal_lattice_vol(const double&, const double&, const double&, const double&, const double&, const double&); // in which you can use the function defined above that converting abc_angles to veca, vecb, vecc and call the function with the same name as this.

// then functions exporting cif.
static void write_cif_nosym(const std::string& fcif, const double& a, const double&b, const double&c, const double& alpha, const double& beta, const double& gamma, const std::vector<std::string>& labels, const std::vector<int>& natoms, const std::vector<ModuleBase::Vector3<double>>& taud, const std::vector<double>& occ = {}); // the most general, because the symmetry information may sometimes be inaccessible. Then you can find this function is even not too worthy to be tested, yes it should be like that.
static void write_cif_sym(...); // more specific, can write the "symmetrized" cif

// then you can add some workflow functions like
static void to_cif(const UnitCell& ucell, const bool with_sym = false); // you parse information here and pass to functions above
// even more but optionally, you have output of cif, it is possible to support an input from cif, but it may need much more design
```
Thanks very much